nutils/pointsseq.py

```diff
@@ -0,0 +1,648 @@
+    '''Return ``self+other``.'''
+
+    if isinstance(other, PointsSequence):
+      return self.chain(other)
+    else:
+      return NotImplemented
+
+  def __mul__(self, other):
+    '''Return ``self*other``.'''
+
+    if numeric.isint(other):
+      return self.repeat(other)
+    elif isinstance(other, PointsSequence):
+      return self.product(other)
+    else:
+      return NotImplemented
+
+  @abc.abstractmethod
+  def get(self, index):
+    '''Return the points at ``index``.
+
+    Parameters
+    ----------
+    index : :class:`int`
+
+    Returns
+    -------
+    points: :class:`~nutils.points.Points`
+        The points at ``index``.
+    '''
+
+    raise NotImplementedError
+
+  def take(self, indices):
+    '''Return a selection of this sequence.
+
+    Parameters
+    ----------
+    indices : :class:`numpy.ndarray`, ndim: 1, dtype: int
+        The indices of points of this sequence to select.
+
+    Returns
+    -------
+    points: :class:`PointsSequence`
+        The sequence of selected points.
+    '''
+
+    _checktake(len(self), indices)
+    if len(indices) == 0:
+      return Empty(self.ndims)
+    elif len(indices) == 1:
+      return Uniform(self.get(indices[0]), 1)
+    else:
+      return Take(self, indices)
+
+  def compress(self, mask):
+    '''Return a selection of this sequence.
+
+    Parameters
+    ----------
+    mask : :class:`numpy.ndarray`, ndim: 1, dtype: bool
+        A boolean mask of points of this sequence to select.
+
+    Returns
+    -------
+    sequence: :class:`PointsSequence`
+        The sequence of selected points.
+    '''
+
+    _checkcompress(len(self), mask)
+    return self.take(numpy.nonzero(mask)[0])
+
+  def repeat(self, count):
+    '''Return this sequence repeated ``count`` times.
+
+    Parameters
+    ----------
+    count : :class:`int`
+
+    Returns
+    -------
+    sequence : :class:`PointsSequence`
+        This sequence repeated ``count`` times.
+    '''
+
+    if count == 0:
+      return Empty(self.ndims)
+    elif count == 1:
+      return self
+    else:
+      return Repeat(self, count)
+
+  def product(self, other):
+    '''Return the product of this sequence with another sequence.
+
+    Parameters
+    ----------
+    other : :class:`PointsSequence`
+
+    Returns
+    -------
+    sequence : :class:`PointsSequence`
+        This product sequence.
+    '''
+
+    return Product(self, other)
+
+  def chain(self, other):
+    '''Return the chained sequence of this sequence with ``other``.
+
+    Parameters
+    ----------
+    other : :class:`PointsSequence`
+
+    Returns
+    -------
+    sequence : :class:`PointsSequence`
+        The chained sequence.
+    '''
+
+    if not other:
+      return self
+    elif not self:
+      return other
+    elif self == other:
+      return Repeat(self, 2)
+    else:
+      return Chain(self, other).balanced
+
+  @property
+  def isuniform(self):
+    '''``True`` if all points in this sequence are equal.'''
```

Comment:
since we expect all sequences for which isuniform is True to be Uniform (otherwise they're anyhow inefficient) can't we just do isinstance checks from within Uniform instead of having this property? just a thought; maybe the main reason for having this is the inefficiencywarnings.

Reply:
With the private implementations the insinstance check is not possible anymore.